addons/sale_timesheet/models/account.py

```diff
@@ -55,12 +70,64 @@ def _get_timesheet_cost(self, values):
             amount = -unit_amount * cost
             amount_converted = employee.currency_id.compute(amount, analytic_account.currency_id)
 
-            return {
+            result.update({
                 'amount': amount_converted,
                 'product_uom_id': uom.id,
                 'account_id': account_id,
-            }
-        return {}
+            })
+        return result
+
+    def _get_timesheet_billing_values(self, values):
+        """
+            If invoice on delivered quantity:
+                timesheet hours * (SO Line Price) * (1- discount),
+            elif invoice on ordered quantities & create task:
+                min (
+                    timesheet hours * (SO Line unit price) * (1- discount),
+                    TOTAL SO - TOTAL INVOICED - sum(timesheet revenues with invoice_id=False)
+                )
+            else:
+                0
+        """
+        result = {}
+        if self.env.context.get('create'):  # avoid bad loop
+            return result
+
+        unit_amount = values.get('unit_amount', 0.0) or self.unit_amount
+        billable_type = 'non_billable'
+        revenue = 0.0
+
+        # set the revenue and billable type according to the product and the SO line
+        so_line_id = values.get('so_line') or self.so_line.id
+        so_line = self.env['sale.order.line'].browse(so_line_id) if so_line_id else self.env['sale.order.line'].browse()
+        if so_line.product_id.type == 'service':
+            # find the analytic account to convert revenue into its currency
+            account_id = values.get('account_id') or self.account_id.id
+            analytic_account = self.env['account.analytic.account'].browse(account_id)
+            # calculate the revenue on the timesheet
+            if so_line.product_id.invoice_policy == 'delivery':
+                delivered_revenue = unit_amount * so_line.price_unit * (1-so_line.discount)
```

Comment:
Extract delivery_revenue as it is computed two times. Moreover it helps understanding the algorithm that the second point is the minimum of the first point and what remains to invoice.